pkg/ip/ip.go

```diff
@@ -0,0 +1,219 @@
+	sort.Sort(ByMask(removeCIDRs))
+
+PreLoop:
+	// Remove CIDRs which are contained within CIDRs that we want to remove;
+	// such CIDRs are redundant.
+	for j, removeCIDR := range removeCIDRs {
+		for i, removeCIDR2 := range removeCIDRs {
+			if i == j {
+				continue
+			}
+			if removeCIDR.Contains(removeCIDR2.IP) {
+				removeCIDRs = append(removeCIDRs[:i], removeCIDRs[i+1:]...)
+				// Re-trigger loop since we have modified the slice we are iterating over.
+				goto PreLoop
+			}
+		}
+	}
+
+Loop:
+	for j, remove := range removeCIDRs {
+		for i, allowCIDR := range allowCIDRs {
+
+			// Don't allow comparison of different address spaces.
+			if !govalidator.IsIPv4(allowCIDR.IP.String()) && govalidator.IsIPv4(remove.IP.String()) ||
+				govalidator.IsIPv4(allowCIDR.IP.String()) && !govalidator.IsIPv4(remove.IP.String()) {
+				return nil, fmt.Errorf("cannot compare different address spaces")
+			}
+
+			// Only remove CIDR if it is contained in the subnet we are allowing.
+			if allowCIDR.Contains(remove.IP) {
+				nets, err := removeCIDR(allowCIDR, remove)
+				if err != nil {
+					return nil, err
+				}
+
+				// Remove CIDR that we have just processed and append new CIDRs
+				// that we computed from removing the CIDR to remove.
+				allowCIDRs = append(allowCIDRs[:i], allowCIDRs[i+1:]...)
+				allowCIDRs = append(allowCIDRs, nets...)
+				goto Loop
+			} else if remove.Contains(allowCIDR.IP.Mask(allowCIDR.Mask)) {
+				// If a CIDR that we want to remove contains a CIDR in the list
+				// that is allowed, then we can just remove the CIDR to allow.
+				allowCIDRs = append(allowCIDRs[:i], allowCIDRs[i+1:]...)
+				goto Loop
+			}
+		}
+		removeCIDRs = append(removeCIDRs[:j], removeCIDRs[j+1:]...)
+		goto Loop
+	}
+
+	return allowCIDRs, nil
+}
+
+func getFirstIP(ipNet *net.IPNet) *net.IP {
+	var mask net.IP
+
+	if govalidator.IsIPv6(ipNet.IP.String()) {
+		mask = make(net.IP, net.IPv6len)
+		for i := 0; i < len(ipNet.Mask); i++ {
+			mask[net.IPv6len-i-1] = ipNet.IP[net.IPv6len-i-1] & ^ipNet.Mask[i]
+		}
+	} else {
+		mask = make(net.IP, net.IPv4len)
+		for i := 0; i < net.IPv4len; i++ {
+			mask[net.IPv4len-i-1] = ipNet.IP[net.IPv6len-i-1] & ^ipNet.Mask[i]
+		}
+	}
+
+	return &mask
+}
+
+func removeCIDR(allowCIDR, removeCIDR *net.IPNet) ([]*net.IPNet, error) {
+	var allows []*net.IPNet
+	var allowIsIpv4, removeIsIpv4 bool
+	var allowBitLen int
+
+	if govalidator.IsIPv4(allowCIDR.IP.String()) {
+		allowIsIpv4 = true
+		allowBitLen = ipv4BitLen
+	} else {
+		allowBitLen = ipv6BitLen
+	}
+
+	if govalidator.IsIPv4(removeCIDR.IP.String()) {
+		removeIsIpv4 = true
+	}
+
+	if (removeIsIpv4 && !allowIsIpv4) || (!removeIsIpv4 && allowIsIpv4) {
+		return nil, fmt.Errorf("cannot mix different address spaces")
+	}
+
+	// Get size of each CIDR mask.
+	allowSize, _ := allowCIDR.Mask.Size()
+	removeSize, _ := removeCIDR.Mask.Size()
+
+	if allowSize >= removeSize {
+		return nil, fmt.Errorf("allow CIDR must be a superset of remove CIDR")
```

Comment:
This can be done in `Validate()` so you don't need to error out here.

Reply:
Quoting above comment:
> This is in its own package, though. If we want to use these functions elsewhere for whatever reason, it is useful to have an error returned.